pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,353 @@
+		bfs.log.Error("Invalid mode specified", zap.String("mode", bfs.cfg.Mode))
+		close(bfs.quit)
+		return
+	}
+
+	if err != nil {
+		bfs.log.Error("Fetch operation failed", zap.Error(err))
+		close(bfs.quit)
+		return
+	}
+}
+
+func (bfs *Service) GetHeaders() error {
+	startIndex := bfs.chain.HeaderHeight() / 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter("index_header", fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter("index_header", fmt.Sprintf("%d", startIndex), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			headerOids, err := bfs.search(prm)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find 'index_header' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(headerOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No 'index_header' object found with index %d, stopping.", startIndex))
+				return errors.New("no 'index_header' object found")
+			}
+
+			data, err := bfs.get(headerOids[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch 'index_header' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			err = bfs.processHeaders(data)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to process headers for index %d", startIndex), zap.Error(err))
+				return err
+			}
+			startIndex++
+		}
+	}
+}
+
+// Shutdown implements the core.Service interface.
+func (bfs *Service) Shutdown() {
+	bfs.log.Info("Shutting down Block Fetcher Service")
+	close(bfs.quit)
+}
+
+// IsActive implements the Service interface.
+func (bfs *Service) IsActive() bool {
+	if !bfs.cfg.Enabled {
+		return false
+	}
+	select {
+	case <-bfs.quit:
+		return false
+	default:
+		return true
+	}
+}
+
+func (bfs *Service) fetchData() error {
+	startIndex := bfs.chain.BlockHeight()
+	batchSize := uint32(30)
+
+	for {
+		select {
+		case <-bfs.quit:
+			bfs.log.Info("Stopping fetchData operation.")
+			return nil
+		default:
+			bfs.log.Info(fmt.Sprintf("Current BlockHeight %d", bfs.chain.BlockHeight()))
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter("index_block_2", fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter("index_block_2", fmt.Sprintf("%d", startIndex+batchSize-1), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			blockOids, err := bfs.search(prm)
+			bfs.log.Info(fmt.Sprintf("Found %d blocks from index %d to %d", len(blockOids), startIndex, startIndex+batchSize-1))
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find block with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No block found with index %d, stopping.", startIndex))
+				return errors.New("no block found")
+			}
+			for _, oidBlock := range blockOids {
+				go func(oidBlock oid.ID) {
```

Comment:
You need to wait outside this `for` loop until downloading of this batch is finished, use wait group for it. And only after that you may go to the next batch. Likely it's the reason of your "context deadline exceeded" error.